tools/build_tar/buildtar.go

```diff
@@ -330,7 +329,105 @@ func (f *tarFile) addTar(toAdd string) error {
 }
 
 func (f *tarFile) addDeb(toAdd string) error {
-	return fmt.Errorf("addDeb unimplemented")
+	pkgDataFound, pkgMetadataFound := false, false
+	var r io.Reader
+	file, err := os.Open(toAdd)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	r = file
+	r = bufio.NewReader(r)
+	tr := ar.NewReader(r)
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		parts := strings.SplitN(header.Name, ".", 2)
+		name := strings.TrimRight(header.Name, "/")
+		switch parts[0] {
+		case "data":
+			pkgDataFound = true
+			err = f.addTar(tr, name)
+			if err != nil {
+				return err
+			}
+		case "control":
+			pkgMetadataFound = true
+			f.addPkgMetadata(tr, name)
+		}
+	}
+	if !pkgDataFound {
+		return fmt.Errorf("%s does not contain a data file", toAdd)
+	}
+	if !pkgMetadataFound {
+		return fmt.Errorf("%s does not contain a control file", toAdd)
+	}
+	return nil
+}
+
+func (f *tarFile) addPkgMetadata(r io.Reader, toAdd string) error {
+	r, err := decompress(r, toAdd)
+	if err != nil {
+		return err
+	}
+	tr := tar.NewReader(r)
+	controlFileFound := false
+	var buf bytes.Buffer
+	var header *tar.Header
+	for {
+		header, err = tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if filepath.Base(header.Name) == pkgMetadataFile {
+			controlFileFound = true
+			io.Copy(&buf, r)
+			break
+		}
+
+	}
+	if !controlFileFound {
+		return fmt.Errorf("%s does not contain control file", toAdd)
+	}
+	metadata := buf.String()
+	destinationFile := filepath.Join(dpkgStatusDir, parsePkgName(metadata, toAdd))
+	header.Name = destinationFile
```

Comment:
combine the previous lines with this one - `header.Name = filepath.Join(..., parsePkgName(metadata.String(), toAdd))`.

also, you could make `parsePkgName` take a `[]byte` and use `FindSubmatch` instead of `FindStringSubmatch` to skip some unnecessary string conversion.

Reply:
done